rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,239 @@
+      },
+      term.persistent
+    )
+
+  def substitute(term: Eval, env: Env): Eval =
+    substitute(term.channel, env) match {
+      case Left(_channel) => Eval(_channel)
+      case Right(par)     => throw new Error(s"Cannot dereference process: $par")
+    }
+
+  def rename(term: Eval, j: Int): Eval =
+    Eval(rename(term.channel, j))
+
+  def substitute(term: New, env: Env): New =
+    New(term.count, substitute(term.p, env))
+
+  def rename(term: New, j: Int): New =
+    New(term.count, rename(term.p, j))
+
+  def substitute(term: Receive, env: Env): Receive =
+    Receive(
+      for { (xs, chan) <- term.binds } yield {
+        substitute(chan, env) match {
+          case Left(_chan) => (xs, _chan)
+          case Right(par)  => throw new Error(s"Cannot receive on expression: $par")
+        }
+      },
+      substitute(term.body, env),
+      term.persistent,
+      term.count
+    )
+
+  def rename(term: Receive, j: Int): Receive =
+    Receive(
+      for ((xs, channel) <- term.binds) yield {
+        (xs, rename(channel, j))
+      },
+      rename(term.body, j),
+      term.persistent,
+      term.count
+    )
+
+  def substitute(term: Var, env: Env): Either[Var, Data] =
+    term match {
+      case BoundVar(level) =>
+        env.get(level) match {
+          case Some(Left(quote)) => Right(Left(rename(quote, env.level)))
+          case Some(Right(par))  => Right(Right(rename(par, env.level)))
+          case None =>
+            if (level - env.level < 0) Left(BoundVar(0))
+            else Left(BoundVar(level - env.level))
+        }
+      case _ => Left(term)
+    }
+
+  def rename(term: Var, j: Int): Var =
+    term match {
+      case BoundVar(i) =>
+        if (i + j < 0) BoundVar(0)
+        else BoundVar(i + j)
+      case _ => term
+    }
+
+  def rename(term: Expr, j: Int): Expr =
+    term match {
+      case EVar(v: Var)             => EVar(rename(v, j))
+      case ENot(p: Par)             => ENot(rename(p, j))
+      case ENeg(p: Par)             => ENeg(rename(p, j))
+      case EMult(p1: Par, p2: Par)  => EMult(rename(p1, j), rename(p2, j))
+      case EDiv(p1: Par, p2: Par)   => EDiv(rename(p1, j), rename(p2, j))
+      case EPlus(p1: Par, p2: Par)  => EPlus(rename(p1, j), rename(p2, j))
+      case EMinus(p1: Par, p2: Par) => EMinus(rename(p1, j), rename(p2, j))
+      case ELt(p1: Par, p2: Par)    => ELt(rename(p1, j), rename(p2, j))
+      case ELte(p1: Par, p2: Par)   => ELte(rename(p1, j), rename(p2, j))
+      case EGt(p1: Par, p2: Par)    => EGt(rename(p1, j), rename(p2, j))
+      case EGte(p1: Par, p2: Par)   => EGte(rename(p1, j), rename(p2, j))
+      case EEq(p1: Par, p2: Par)    => EEq(rename(p1, j), rename(p2, j))
+      case ENeq(p1: Par, p2: Par)   => ENeq(rename(p1, j), rename(p2, j))
+      case EAnd(p1: Par, p2: Par)   => EAnd(rename(p1, j), rename(p2, j))
+      case EOr(p1: Par, p2: Par)    => EOr(rename(p1, j), rename(p2, j))
+    }
+
+  def rename(exp: Ground, j: Int): Ground =
```

Comment:
Does scala really allow this? A Ground is an expression. Shouldn't this conflict with the above?

Even if it does, I'd much rather you split them off in the function above and had a function with a different name for grounds.

Reply:
On the JVM, when the argument is resolved, it's resolved to the most specific type in the hierarchy, so when `rename(GBool(true),i)` is called, `GBool(true)` resolves to `Ground` not `Expr`, thus the second rename is called. 

I can totally split them off, but I'm curious why?

Reply:
It's fine. I don't like code where I have to know odd pieces of the JVM in order to understand them.